### PR TITLE
Add IBC to Rho

### PR DESCRIPTION
**Pull request by okwme**
Added IBC and ICA accounts plus E2E tests

### CODE REVIEW

**Comment by glnro** [comment]
Building this and running a chain locally: 
```
./gaiad start --minimum-gas-prices "0.002uatom" --home ~/.rho_ibc
10:15PM INF starting node with ABCI Tendermint in-process
10:15PM INF starting service impl=multiAppConn module=proxy service=multiAppConn
10:15PM INF starting service connection=query impl=localClient module=abci-client service=localClient
10:15PM INF starting service connection=snapshot impl=localClient module=abci-client service=localClient
10:15PM INF starting service connection=mempool impl=localClient module=abci-client service=localClient
10:15PM INF starting service connection=consensus impl=localClient module=abci-client service=localClient
10:15PM INF starting service impl=EventBus module=events service=EventBus
10:15PM INF starting service impl=PubSub module=pubsub service=PubSub
10:15PM INF starting service impl=IndexerService module=txindex service=IndexerService
10:15PM INF ABCI Handshake App Info hash= height=0 module=consensus protocol-version=0 software-version=rho-ibc-fa93bb8ca1cefcb658ccefd063a7bac7af82d45e
10:15PM INF ABCI Replay Blocks appHeight=0 module=consensus stateHeight=0 storeHeight=0
10:15PM INF initializing blockchain state from genesis.json
10:15PM INF asserting crisis invariants inv=0/12 module=x/crisis name=gov/module-account
10:15PM INF asserting crisis invariants inv=1/12 module=x/crisis name=distribution/nonnegative-outstanding
10:15PM INF asserting crisis invariants inv=2/12 module=x/crisis name=distribution/can-withdraw
10:15PM INF asserting crisis invariants inv=3/12 module=x/crisis name=distribution/reference-count
10:15PM INF asserting crisis invariants inv=4/12 module=x/crisis name=distribution/module-account
10:15PM INF asserting crisis invariants inv=5/12 module=x/crisis name=staking/module-accounts
10:15PM INF asserting crisis invariants inv=6/12 module=x/crisis name=staking/nonnegative-power
10:15PM INF asserting crisis invariants inv=7/12 module=x/crisis name=staking/positive-delegation
10:15PM INF asserting crisis invariants inv=8/12 module=x/crisis name=staking/delegator-shares
10:15PM INF asserting crisis invariants inv=9/12 module=x/crisis name=bank/nonnegative-outstanding
10:15PM INF asserting crisis invariants inv=10/12 module=x/crisis name=bank/total-supply
10:15PM INF asserting crisis invariants inv=11/12 module=x/crisis name=group/Group-TotalWeight
10:15PM INF asserted all invariants duration=0.758875 height=0 module=x/crisis
panic: kv store with key <nil> has not been registered in stores

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/store/cachemulti.Store.GetKVStore({{0x1056cb2b8, 0x140002c4300}, 0x14000a69260, 0x14000c0a090, {0x0, 0x0}, 0x0, 0x14000c0a0c0}, {0x10565b708, 0x0})
	github.com/cosmos/cosmos-sdk@v0.46.0-beta2/store/cachemulti/store.go:185 +0x10c
github.com/cosmos/cosmos-sdk/types.Context.KVStore({{0x10569b398, 0x14000126008}, {0x1056d1dc0, 0x140002c4780}, {{0x0, 0x0}, {0x14000c04929, 0x5}, 0x0, {0x306c8ba0, ...}, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.46.0-beta2/types/context.go:253 +0x64
github.com/cosmos/ibc-go/v3/modules/apps/transfer/keeper.Keeper.SetPort({{0x10565b708, 0x0}, {0x1056cb248, 0x1400019d8a0}, {{0x1056cb248, 0x1400019d8a0}, 0x140003c0670, {0x10565b708, 0x1400050b9b0}, {0x10565b758, ...}, ...}, ...}, ...)
	github.com/cosmos/ibc-go/v3@v3.0.0/modules/apps/transfer/keeper/keeper.go:94 +0x54
github.com/cosmos/ibc-go/v3/modules/apps/transfer/keeper.Keeper.InitGenesis({{0x10565b708, 0x0}, {0x1056cb248, 0x1400019d8a0}, {{0x1056cb248, 0x1400019d8a0}, 0x140003c0670, {0x10565b708, 0x1400050b9b0}, {0x10565b758, ...}, ...}, ...}, ...)
	github.com/cosmos/ibc-go/v3@v3.0.0/modules/apps/transfer/keeper/genesis.go:13 +0x70
github.com/cosmos/ibc-go/v3/modules/apps/transfer.AppModule.InitGenesis({{}, {{0x10565b708, 0x0}, {0x1056cb248, 0x1400019d8a0}, {{0x1056cb248, 0x1400019d8a0}, 0x140003c0670, {0x10565b708, 0x1400050b9b0}, ...}, ...}}, ...)
	github.com/cosmos/ibc-go/v3@v3.0.0/modules/apps/transfer/module.go:130 +0xd0
github.com/cosmos/cosmos-sdk/types/module.(*Manager).InitGenesis(0x140003d8480, {{0x10569b398, 0x14000126008}, {0x1056d1dc0, 0x140002c4780}, {{0x0, 0x0}, {0x14000c04929, 0x5}, 0x0, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.46.0-beta2/types/module/module.go:324 +0x3a8
github.com/cosmos/gaia/v8/app.(*GaiaApp).InitChainer(0x1400071a000, {{0x10569b398, 0x14000126008}, {0x1056d1dc0, 0x140002c4780}, {{0x0, 0x0}, {0x14000c04929, 0x5}, 0x0, ...}, ...}, ...)
	github.com/cosmos/gaia/v8/app/app.go:815 +0x178
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).InitChain(0x140008756c0, {{0x306c8ba0, 0xeda1355f2, 0x0}, {0x14000c04929, 0x5}, 0x140002b2f60, {0x106928a60, 0x0, 0x0}, ...})
	github.com/cosmos/cosmos-sdk@v0.46.0-beta2/baseapp/abci.go:61 +0x2e0
github.com/tendermint/tendermint/abci/client.(*localClient).InitChainSync(0x140004030e0, {0x10569b398, 0x14000126008}, {{0x306c8ba0, 0xeda1355f2, 0x0}, {0x14000c04929, 0x5}, 0x140002b2f60, {0x106928a60, ...}, ...})
	github.com/tendermint/tendermint@v0.35.2/abci/client/local_client.go:273 +0x110
github.com/tendermint/tendermint/internal/proxy.(*appConnConsensus).InitChainSync(0x140003ab428, {0x10569b398, 0x14000126008}, {{0x306c8ba0, 0xeda1355f2, 0x0}, {0x14000c04929, 0x5}, 0x140002b2f60, {0x106928a60, ...}, ...})
	github.com/tendermint/tendermint@v0.35.2/internal/proxy/app_conn.go:85 +0x11c
github.com/tendermint/tendermint/internal/consensus.(*Handshaker).ReplayBlocks(0x14000b25748, {{{0xb, 0x0}, {0x10478b4b3, 0x11}}, {0x14000c04929, 0x5}, 0x1, 0x0, {{0x0, ...}, ...}, ...}, ...)
	github.com/tendermint/tendermint@v0.35.2/internal/consensus/replay.go:320 +0xd18
github.com/tendermint/tendermint/internal/consensus.(*Handshaker).Handshake(0x14000b25748, {0x1056d3510, 0x14000618820})
	github.com/tendermint/tendermint@v0.35.2/internal/consensus/replay.go:269 +0x400
github.com/tendermint/tendermint/node.doHandshake({0x1056c9178, 0x14000a64230}, {{{0xb, 0x0}, {0x10478b4b3, 0x11}}, {0x14000c04929, 0x5}, 0x1, 0x0, ...}, ...)
	github.com/tendermint/tendermint@v0.35.2/node/setup.go:117 +0x210
github.com/tendermint/tendermint/node.makeNode(0x140008f1200, {0x10566dd48, 0x14000c18b40}, {{0x1400066f980, 0x28}, {0x1056a45f8, 0x14000c01b60}}, 0x140002b2da0, 0x140005f45f8, 0x105615028, ...)
	github.com/tendermint/tendermint@v0.35.2/node/node.go:215 +0x7f8
github.com/tendermint/tendermint/node.New(0x140008f1200, {0x10569c2b0, 0x14000bca5a0}, 0x140002b2da0, 0x140003d9480)
	github.com/tendermint/tendermint@v0.35.2/node/public.go:54 +0x2ec
github.com/cosmos/cosmos-sdk/server.startInProcess(0x1400001fd80, {{0x0, 0x0, 0x0}, {0x1056e8878, 0x1400098fdd0}, 0x0, {0x0, 0x0}, {0x1056dd6d8, ...}, ...}, ...)
	github.com/cosmos/cosmos-sdk@v0.46.0-beta2/server/start.go:286 +0x5d4
github.com/cosmos/cosmos-sdk/server.StartCmd.func2(0x140003dc280, {0x140002c4bc0, 0x0, 0x4})
	github.com/cosmos/cosmos-sdk@v0.46.0-beta2/server/start.go:137 +0x10c
github.com/spf13/cobra.(*Command).execute(0x140003dc280, {0x140002c4b80, 0x4, 0x4})
	github.com/spf13/cobra@v1.4.0/command.go:856 +0x668
github.com/spf13/cobra.(*Command).ExecuteC(0x14000050c80)
	github.com/spf13/cobra@v1.4.0/command.go:974 +0x410
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.4.0/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	github.com/spf13/cobra@v1.4.0/command.go:895
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0x14000050c80, {0x0, 0x0}, {0x140003c5e80, 0xf})
	github.com/cosmos/cosmos-sdk@v0.46.0-beta2/server/cmd/execute.go:36 +0x244
main.main()
	github.com/cosmos/gaia/v8/cmd/gaiad/main.go:16 +0x40
```

**Comment by yaruwangway** [comment]
@glnro , i have no problem to run

**Review by yaruwangway — APPROVED** [review]

File: ante/ante.go [hunk 1]
```diff
@@ -1,55 +1,55 @@
...
+// func NewAnteHandler(opts HandlerOptions) (sdk.AnteHandler, error) {
+// 	if opts.AccountKeeper == nil {
+// 		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "account keeper is required for AnteHandler")
+// 	}
+// 	if opts.BankKeeper == nil {
+// 		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "bank keeper is required for AnteHandler")
+// 	}
+// 	if opts.SignModeHandler == nil {
+// 		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "sign mode handler is required for ante builder")
+// 	}
```
Comment (yaruwangway) [inline ante/ante.go:29]: just for later when adding back ante handler 
need to add `
	if options.IBCKeeper == nil {
		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "IBC keeper is required for middlewares")
	}
`

**Comment by faddat** [comment]
use this, and end the suffering, and please, help me to fix the last test:

https://github.com/cosmos/ibc-go/pull/1413

File: app/const.go [hunk 2]
```diff
@@ -2,32 +2,32 @@ package gaia
...
-	// stakingMsgUndelegate                = "/cosmos.staking.v1beta1.MsgUndelegate"
-	// stakingMsgBeginRedelegate           = "/cosmos.staking.v1beta1.MsgBeginRedelegate"
-	// stakingMsgCreateValidator           = "/cosmos.staking.v1beta1.MsgCreateValidator"
-	// vestingMsgCreateVestingAccount      = "/cosmos.vesting.v1beta1.MsgCreateVestingAccount"
-	// transferMsgTransfer                 = "/ibc.applications.transfer.v1.MsgTransfer"
-	// liquidityMsgCreatePool              = "/tendermint.liquidity.v1beta1.MsgCreatePool"
-	// liquidityMsgSwapWithinBatch         = "/tendermint.liquidity.v1beta1.MsgSwapWithinBatch"
-	// liquidityMsgDepositWithinBatch      = "/tendermint.liquidity.v1beta1.MsgDepositWithinBatch"
-	// liquidityMsgWithdrawWithinBatch     = "/tendermint.liquidity.v1beta1.MsgWithdrawWithinBatch"
+	authzMsgExec                        = "/cosmos.authz.v1beta1.MsgExec"
```
Comment (tac0turtle) [inline app/const.go:7]: what do these constants do? (this is slightly concerning to me)

**Pull request merged**